exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/FunctionImplementationRegistry.java

```diff
@@ -138,18 +171,25 @@ private String functionReplacement(FunctionCall functionCall) {
 
   /**
    * Find the Drill function implementation that matches the name, arg types and return type.
-   * @param name
-   * @param argTypes
-   * @param returnType
-   * @return
+   * If exact function implementation was not found and in case if Dynamic UDF Support is enabled
+   * loads all missing remote functions and tries to find Drill implementation one more time.
    */
   public DrillFuncHolder findExactMatchingDrillFunction(String name, List<MajorType> argTypes, MajorType returnType) {
+    return findExactMatchingDrillFunction(name, argTypes, returnType, true);
+  }
+
+  private DrillFuncHolder findExactMatchingDrillFunction(String name, List<MajorType> argTypes, MajorType returnType, boolean retry) {
     for (DrillFuncHolder h : drillFuncRegistry.getMethods(name)) {
       if (h.matches(returnType, argTypes)) {
         return h;
       }
     }
 
+    if (retry && optionManager != null && optionManager.getOption(ExecConstants.DYNAMIC_UDF_SUPPORT_ENABLED).bool_val) {
+      if (loadRemoteFunctions()) {
```

Comment:
Race condition. Two fragments each call into this method to find functions. Both don't find the function. Both will get the remote functions simutaneously. Should there be some synchronization so that the remote checks are either serialized, or we limit the number of resyncs per unit time?